<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ midasHLA</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">midasHLA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/MiDAS_tutorial.html">MiDAS tutorial</a>
    </li>
    <li>
      <a href="../articles/MiDAS_vignette.html">MiDAS quick start</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="HWETest.html">HWETest()</a></code> </p>
        </td>
        <td><p>Test for Hardy Weinberg equilibrium</p></td>
      </tr><tr><td>
          <p><code><a href="LRTest.html">LRTest()</a></code> </p>
        </td>
        <td><p>Likelihood ratio test</p></td>
      </tr><tr><td>
          <p><code><a href="MiDAS-class.html">getExperiments(<i>&lt;MiDAS&gt;</i>)</a></code> <code><a href="MiDAS-class.html">getHlaCalls(<i>&lt;MiDAS&gt;</i>)</a></code> <code><a href="MiDAS-class.html">getKirCalls(<i>&lt;MiDAS&gt;</i>)</a></code> <code><a href="MiDAS-class.html">getPlaceholder(<i>&lt;MiDAS&gt;</i>)</a></code> <code><a href="MiDAS-class.html">getOmnibusGroups(<i>&lt;MiDAS&gt;</i>)</a></code> <code><a href="MiDAS-class.html">getFrequencies(<i>&lt;MiDAS&gt;</i>)</a></code> <code><a href="MiDAS-class.html">filterByFrequency(<i>&lt;MiDAS&gt;</i>)</a></code> <code><a href="MiDAS-class.html">filterByOmnibusGroups(<i>&lt;MiDAS&gt;</i>)</a></code> <code><a href="MiDAS-class.html">filterByVariables(<i>&lt;MiDAS&gt;</i>)</a></code> <code><a href="MiDAS-class.html">getAllelesForAA(<i>&lt;MiDAS&gt;</i>)</a></code> </p>
        </td>
        <td><p>MiDAS class</p></td>
      </tr><tr><td>
          <p><code><a href="MiDAS_tut_HLA.html">MiDAS_tut_HLA</a></code> </p>
        </td>
        <td><p>MiDAS tutorial HLA data</p></td>
      </tr><tr><td>
          <p><code><a href="MiDAS_tut_KIR.html">MiDAS_tut_KIR</a></code> </p>
        </td>
        <td><p>MiDAS tutorial KIR data</p></td>
      </tr><tr><td>
          <p><code><a href="MiDAS_tut_object.html">MiDAS_tut_object</a></code> </p>
        </td>
        <td><p>MiDAS tutorial MiDAS object</p></td>
      </tr><tr><td>
          <p><code><a href="MiDAS_tut_pheno.html">MiDAS_tut_pheno</a></code> </p>
        </td>
        <td><p>MiDAS tutorial phenotype data</p></td>
      </tr><tr><td>
          <p><code><a href="aaVariationToCounts.html">aaVariationToCounts()</a></code> </p>
        </td>
        <td><p>Transform amino acid variation data frame into counts table</p></td>
      </tr><tr><td>
          <p><code><a href="adjustPValues.html">adjustPValues()</a></code> </p>
        </td>
        <td><p>Adjust P-values for Multiple Comparisons</p></td>
      </tr><tr><td>
          <p><code><a href="allele_frequencies.html">allele_frequencies</a></code> </p>
        </td>
        <td><p>Alleles frequencies scraped from allelefrequencies.net</p></td>
      </tr><tr><td>
          <p><code><a href="analyzeAssociations.html">analyzeAssociations()</a></code> </p>
        </td>
        <td><p>Association analysis</p></td>
      </tr><tr><td>
          <p><code><a href="analyzeConditionalAssociations.html">analyzeConditionalAssociations()</a></code> </p>
        </td>
        <td><p>Stepwise conditional association analysis</p></td>
      </tr><tr><td>
          <p><code><a href="applyInheritanceModel.html">applyInheritanceModel()</a></code> </p>
        </td>
        <td><p>Apply inheritance model</p></td>
      </tr><tr><td>
          <p><code><a href="as.data.frame.MiDAS.html">as.data.frame(<i>&lt;MiDAS&gt;</i>)</a></code> </p>
        </td>
        <td><p>Coerce MiDAS to Data Frame</p></td>
      </tr><tr><td>
          <p><code><a href="backquote.html">backquote()</a></code> </p>
        </td>
        <td><p>Backquote character</p></td>
      </tr><tr><td>
          <p><code><a href="characterMatches.html">characterMatches()</a></code> </p>
        </td>
        <td><p>Check if character matches one of possible values</p></td>
      </tr><tr><td>
          <p><code><a href="checkAlleleFormat.html">checkAlleleFormat()</a></code> </p>
        </td>
        <td><p>Check HLA allele format</p></td>
      </tr><tr><td>
          <p><code><a href="checkColDataFormat.html">checkColDataFormat()</a></code> </p>
        </td>
        <td><p>Assert colData data</p></td>
      </tr><tr><td>
          <p><code><a href="checkHlaCallsFormat.html">checkHlaCallsFormat()</a></code> </p>
        </td>
        <td><p>Assert hla calls data frame format</p></td>
      </tr><tr><td>
          <p><code><a href="checkKirCallsFormat.html">checkKirCallsFormat()</a></code> </p>
        </td>
        <td><p>Assert KIR counts data frame format</p></td>
      </tr><tr><td>
          <p><code><a href="checkKirGenesFormat.html">checkKirGenesFormat()</a></code> </p>
        </td>
        <td><p>Check KIR genes format</p></td>
      </tr><tr><td>
          <p><code><a href="checkStatisticalModel.html">checkStatisticalModel()</a></code> </p>
        </td>
        <td><p>Assert statistical model</p></td>
      </tr><tr><td>
          <p><code><a href="colnamesMatches.html">colnamesMatches()</a></code> </p>
        </td>
        <td><p>Check column names</p></td>
      </tr><tr><td>
          <p><code><a href="convertAlleleToVariable.html">convertAlleleToVariable()</a></code> </p>
        </td>
        <td><p>Convert allele numbers to additional variables</p></td>
      </tr><tr><td>
          <p><code><a href="countsToVariables.html">countsToVariables()</a></code> </p>
        </td>
        <td><p>Convert counts table to variables</p></td>
      </tr><tr><td>
          <p><code><a href="dfToExperimentMat.html">dfToExperimentMat()</a></code> </p>
        </td>
        <td><p>Helper transform data frame to experiment matrix</p></td>
      </tr><tr><td>
          <p><code><a href="dict_dist_grantham.html">dict_dist_grantham</a></code> </p>
        </td>
        <td><p>Grantham distance</p></td>
      </tr><tr><td>
          <p><code><a href="distGrantham.html">distGrantham()</a></code> </p>
        </td>
        <td><p>Calculate Grantham distance between amino acid sequences</p></td>
      </tr><tr><td>
          <p><code><a href="experimentMatToDf.html">experimentMatToDf()</a></code> </p>
        </td>
        <td><p>Helper transform experiment matrix to data frame</p></td>
      </tr><tr><td>
          <p><code><a href="filterByFrequency.html">filterByFrequency()</a></code> </p>
        </td>
        <td><p>Filter MiDAS object by frequency</p></td>
      </tr><tr><td>
          <p><code><a href="filterByOmnibusGroups.html">filterByOmnibusGroups()</a></code> </p>
        </td>
        <td><p>Filter MiDAS object by omnibus groups</p></td>
      </tr><tr><td>
          <p><code><a href="filterByVariables.html">filterByVariables()</a></code> </p>
        </td>
        <td><p>Filter MiDAS object by features</p></td>
      </tr><tr><td>
          <p><code><a href="filterExperimentByFrequency.html">filterExperimentByFrequency()</a></code> </p>
        </td>
        <td><p>Filter experiment by frequency</p></td>
      </tr><tr><td>
          <p><code><a href="filterExperimentByVariables.html">filterExperimentByVariables()</a></code> </p>
        </td>
        <td><p>Filter experiment by variable</p></td>
      </tr><tr><td>
          <p><code><a href="filterListByElements.html">filterListByElements()</a></code> </p>
        </td>
        <td><p>Filter list by elements</p></td>
      </tr><tr><td>
          <p><code><a href="formatResults.html">formatResults()</a></code> </p>
        </td>
        <td><p>Pretty format statistical analysis results helper</p></td>
      </tr><tr><td>
          <p><code><a href="getAAFrequencies.html">getAAFrequencies()</a></code> </p>
        </td>
        <td><p>Calculate amino acid frequencies</p></td>
      </tr><tr><td>
          <p><code><a href="getAlleleResolution.html">getAlleleResolution()</a></code> </p>
        </td>
        <td><p>Infer HLA allele resolution</p></td>
      </tr><tr><td>
          <p><code><a href="getAllelesForAA.html">getAllelesForAA()</a></code> </p>
        </td>
        <td><p>Get HLA alleles for amino acid position</p></td>
      </tr><tr><td>
          <p><code><a href="getExperimentFrequencies.html">getExperimentFrequencies()</a></code> </p>
        </td>
        <td><p>Calculate experiment's features frequencies</p></td>
      </tr><tr><td>
          <p><code><a href="getExperimentPopulationMultiplicator.html">getExperimentPopulationMultiplicator()</a></code> </p>
        </td>
        <td><p>Get experiment's population multiplicator</p></td>
      </tr><tr><td>
          <p><code><a href="getExperiments.html">getExperiments()</a></code> </p>
        </td>
        <td><p>Get available experiments in MiDAS object.</p></td>
      </tr><tr><td>
          <p><code><a href="getFrequencies.html">getFrequencies()</a></code> </p>
        </td>
        <td><p>Calculate features frequencies for a given experiment in MiDAS object.</p></td>
      </tr><tr><td>
          <p><code><a href="getFrequencyMask.html">getFrequencyMask()</a></code> </p>
        </td>
        <td><p>Helper function for filtering frequency data frame</p></td>
      </tr><tr><td>
          <p><code><a href="getHlaCalls.html">getHlaCalls()</a></code> </p>
        </td>
        <td><p>Get HLA calls from MiDAS object.</p></td>
      </tr><tr><td>
          <p><code><a href="getHlaCallsGenes.html">getHlaCallsGenes()</a></code> </p>
        </td>
        <td><p>Get HLA calls genes</p></td>
      </tr><tr><td>
          <p><code><a href="getHlaFrequencies.html">getHlaFrequencies()</a></code> </p>
        </td>
        <td><p>Calculate HLA allele frequencies</p></td>
      </tr><tr><td>
          <p><code><a href="getHlaKirInteractions.html">getHlaKirInteractions()</a></code> </p>
        </td>
        <td><p>Get HLA - KIR interactions</p></td>
      </tr><tr><td>
          <p><code><a href="getKIRFrequencies.html">getKIRFrequencies()</a></code> </p>
        </td>
        <td><p>Calculate KIR genes frequencies</p></td>
      </tr><tr><td>
          <p><code><a href="getKirCalls.html">getKirCalls()</a></code> </p>
        </td>
        <td><p>Get KIR calls from MiDAS object.</p></td>
      </tr><tr><td>
          <p><code><a href="getObjectDetails.html">getObjectDetails()</a></code> </p>
        </td>
        <td><p>Get attributes of statistical model object</p></td>
      </tr><tr><td>
          <p><code><a href="getOmnibusGroups.html">getOmnibusGroups()</a></code> </p>
        </td>
        <td><p>Get omnibus groups from MiDAS object.</p></td>
      </tr><tr><td>
          <p><code><a href="getPlaceholder.html">getPlaceholder()</a></code> </p>
        </td>
        <td><p>Get placeholder name from MiDAS object.</p></td>
      </tr><tr><td>
          <p><code><a href="getReferenceFrequencies.html">getReferenceFrequencies()</a></code> </p>
        </td>
        <td><p>Helper transforming reference frequencies</p></td>
      </tr><tr><td>
          <p><code><a href="getVariableAAPos.html">getVariableAAPos()</a></code> </p>
        </td>
        <td><p>Find variable positions in sequence alignment</p></td>
      </tr><tr><td>
          <p><code><a href="hasTidyMethod.html">hasTidyMethod()</a></code> </p>
        </td>
        <td><p>Check if tidy method for class exist</p></td>
      </tr><tr><td>
          <p><code><a href="hlaAlignmentGrantham.html">hlaAlignmentGrantham()</a></code> </p>
        </td>
        <td><p>Helper function returning alignment for Grantham distance calculations</p></td>
      </tr><tr><td>
          <p><code><a href="hlaCallsGranthamDistance.html">hlaCallsGranthamDistance()</a></code> </p>
        </td>
        <td><p>Calculate Grantham distance between HLA alleles</p></td>
      </tr><tr><td>
          <p><code><a href="hlaCallsToCounts.html">hlaCallsToCounts()</a></code> </p>
        </td>
        <td><p>Transform HLA calls to counts table</p></td>
      </tr><tr><td>
          <p><code><a href="hlaToAAVariation.html">hlaToAAVariation()</a></code> </p>
        </td>
        <td><p>Generate amino acid variation matrix</p></td>
      </tr><tr><td>
          <p><code><a href="hlaToVariable.html">hlaToVariable()</a></code> </p>
        </td>
        <td><p>Convert HLA calls to variables</p></td>
      </tr><tr><td>
          <p><code><a href="isCharacterOrNULL.html">isCharacterOrNULL()</a></code> </p>
        </td>
        <td><p>Check if object is character vector or NULL</p></td>
      </tr><tr><td>
          <p><code><a href="isClass.html">isClass()</a></code> </p>
        </td>
        <td><p>Check if object is of class x</p></td>
      </tr><tr><td>
          <p><code><a href="isClassOrNULL.html">isClassOrNULL()</a></code> </p>
        </td>
        <td><p>Check if object is of class x or null</p></td>
      </tr><tr><td>
          <p><code><a href="isCountOrNULL.html">isCountOrNULL()</a></code> </p>
        </td>
        <td><p>Check if object is count or NULL</p></td>
      </tr><tr><td>
          <p><code><a href="isCountsOrZeros.html">isCountsOrZeros()</a></code> </p>
        </td>
        <td><p>Check if vector contains only counts or zeros</p></td>
      </tr><tr><td>
          <p><code><a href="isExperimentCountsOrZeros.html">isExperimentCountsOrZeros()</a></code> </p>
        </td>
        <td><p>Check if frequencies can be calculated for an experiment</p></td>
      </tr><tr><td>
          <p><code><a href="isExperimentInheritanceModelApplicable.html">isExperimentInheritanceModelApplicable()</a></code> </p>
        </td>
        <td><p>Check if experiment is inheritance model applicable</p></td>
      </tr><tr><td>
          <p><code><a href="isFlagOrNULL.html">isFlagOrNULL()</a></code> </p>
        </td>
        <td><p>Check if object is flag or NULL</p></td>
      </tr><tr><td>
          <p><code><a href="isNumberOrNULL.html">isNumberOrNULL()</a></code> </p>
        </td>
        <td><p>Check if object is number or NULL</p></td>
      </tr><tr><td>
          <p><code><a href="isStringOrNULL.html">isStringOrNULL()</a></code> </p>
        </td>
        <td><p>Check if object is string or NULL</p></td>
      </tr><tr><td>
          <p><code><a href="isTRUEorFALSE.html">isTRUEorFALSE()</a></code> </p>
        </td>
        <td><p>Check if object is TRUE or FALSE flag</p></td>
      </tr><tr><td>
          <p><code><a href="iterativeLRT.html">iterativeLRT()</a></code> </p>
        </td>
        <td><p>Iterative likelihood ratio test</p></td>
      </tr><tr><td>
          <p><code><a href="iterativeModel.html">iterativeModel()</a></code> </p>
        </td>
        <td><p>Iteratively evaluate model for different variables</p></td>
      </tr><tr><td>
          <p><code><a href="kableResults.html">kableResults()</a></code> </p>
        </td>
        <td><p>Create association analysis results table in HTML or LaTeX</p></td>
      </tr><tr><td>
          <p><code><a href="kir_frequencies.html">kir_frequencies</a></code> </p>
        </td>
        <td><p>KIR genes frequencies scraped from allelefrequencies.net</p></td>
      </tr><tr><td>
          <p><code><a href="lapply_tryCatch.html">lapply_tryCatch()</a></code> </p>
        </td>
        <td><p>lapply with tryCatch routine</p></td>
      </tr><tr><td>
          <p><code><a href="listMiDASDictionaries.html">listMiDASDictionaries()</a></code> </p>
        </td>
        <td><p>List HLA alleles dictionaries</p></td>
      </tr><tr><td>
          <p><code><a href="midasToWide.html">midasToWide()</a></code> </p>
        </td>
        <td><p>Transform MiDAS to wide format data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="objectHasPlaceholder.html">objectHasPlaceholder()</a></code> </p>
        </td>
        <td><p>Check if placeholder is present in object formula</p></td>
      </tr><tr><td>
          <p><code><a href="omnibusTest.html">omnibusTest()</a></code> </p>
        </td>
        <td><p>Omnibus test</p></td>
      </tr><tr><td>
          <p><code><a href="prepareMiDAS.html">prepareMiDAS()</a></code> </p>
        </td>
        <td><p>Construct a MiDAS object</p></td>
      </tr><tr><td>
          <p><code><a href="prepareMiDAS_hla_NK_ligands.html">prepareMiDAS_hla_NK_ligands()</a></code> </p>
        </td>
        <td><p>Prepare MiDAS data on HLA allele's groups level</p></td>
      </tr><tr><td>
          <p><code><a href="prepareMiDAS_hla_aa.html">prepareMiDAS_hla_aa()</a></code> </p>
        </td>
        <td><p>Prepare MiDAS data on HLA amino acid level</p></td>
      </tr><tr><td>
          <p><code><a href="prepareMiDAS_hla_alleles.html">prepareMiDAS_hla_alleles()</a></code> </p>
        </td>
        <td><p>Prepare MiDAS data on HLA allele level</p></td>
      </tr><tr><td>
          <p><code><a href="prepareMiDAS_hla_custom.html">prepareMiDAS_hla_custom()</a></code> </p>
        </td>
        <td><p>Prepare MiDAS data on custom HLA level</p></td>
      </tr><tr><td>
          <p><code><a href="prepareMiDAS_hla_divergence.html">prepareMiDAS_hla_divergence()</a></code> </p>
        </td>
        <td><p>Prepare MiDAS data on HLA divergence level</p></td>
      </tr><tr><td>
          <p><code><a href="prepareMiDAS_hla_g_groups.html">prepareMiDAS_hla_g_groups()</a></code> </p>
        </td>
        <td><p>Prepare MiDAS data on HLA allele's G groups level</p></td>
      </tr><tr><td>
          <p><code><a href="prepareMiDAS_hla_het.html">prepareMiDAS_hla_het()</a></code> </p>
        </td>
        <td><p>Prepare MiDAS data on HLA heterozygosity level</p></td>
      </tr><tr><td>
          <p><code><a href="prepareMiDAS_hla_kir_interactions.html">prepareMiDAS_hla_kir_interactions()</a></code> </p>
        </td>
        <td><p>Prepare MiDAS data on HLA - KIR interactions level</p></td>
      </tr><tr><td>
          <p><code><a href="prepareMiDAS_hla_supertypes.html">prepareMiDAS_hla_supertypes()</a></code> </p>
        </td>
        <td><p>Prepare MiDAS data on HLA allele's supertypes level</p></td>
      </tr><tr><td>
          <p><code><a href="prepareMiDAS_kir_custom.html">prepareMiDAS_kir_custom()</a></code> </p>
        </td>
        <td><p>Prepare MiDAS data on custom KIR level</p></td>
      </tr><tr><td>
          <p><code><a href="prepareMiDAS_kir_genes.html">prepareMiDAS_kir_genes()</a></code> </p>
        </td>
        <td><p>Prepare MiDAS data on KIR genes level</p></td>
      </tr><tr><td>
          <p><code><a href="prepareMiDAS_kir_haplotypes.html">prepareMiDAS_kir_haplotypes()</a></code> </p>
        </td>
        <td><p>Prepare MiDAS data on KIR haplotypes level</p></td>
      </tr><tr><td>
          <p><code><a href="readHlaAlignments.html">readHlaAlignments()</a></code> </p>
        </td>
        <td><p>Read HLA allele alignments</p></td>
      </tr><tr><td>
          <p><code><a href="readHlaCalls.html">readHlaCalls()</a></code> </p>
        </td>
        <td><p>Read HLA allele calls</p></td>
      </tr><tr><td>
          <p><code><a href="readKirCalls.html">readKirCalls()</a></code> </p>
        </td>
        <td><p>Read KIR calls</p></td>
      </tr><tr><td>
          <p><code><a href="reduceAlleleResolution.html">reduceAlleleResolution()</a></code> </p>
        </td>
        <td><p>Reduce HLA alleles</p></td>
      </tr><tr><td>
          <p><code><a href="reduceHlaCalls.html">reduceHlaCalls()</a></code> </p>
        </td>
        <td><p>Reduce HLA calls resolution</p></td>
      </tr><tr><td>
          <p><code><a href="runMiDAS.html">runMiDAS()</a></code> </p>
        </td>
        <td><p>Run MiDAS statistical analysis</p></td>
      </tr><tr><td>
          <p><code><a href="runMiDASGetVarsFreq.html">runMiDASGetVarsFreq()</a></code> </p>
        </td>
        <td><p>Get variables frequencies from MiDAS</p></td>
      </tr><tr><td>
          <p><code><a href="stringMatches.html">stringMatches()</a></code> </p>
        </td>
        <td><p>Check if string matches one of possible values</p></td>
      </tr><tr><td>
          <p><code><a href="summariseAAPosition.html">summariseAAPosition()</a></code> </p>
        </td>
        <td><p>Summarize amino acid position</p></td>
      </tr><tr><td>
          <p><code><a href="updateModel.html">updateModel()</a></code> </p>
        </td>
        <td><p>Extend and Re-fit a Model Call</p></td>
      </tr><tr><td>
          <p><code><a href="validateFrequencyCutoffs.html">validateFrequencyCutoffs()</a></code> </p>
        </td>
        <td><p>Validate frequency cutoffs</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Christian Hammer, Maciej MigdaÅ‚.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

