<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Construct a MiDAS object — prepareMiDAS • midasHLA</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Construct a MiDAS object — prepareMiDAS" />
<meta property="og:description" content="prepareMiDAS transform HLA alleles calls and KIR calls according
to selected experiments creating a MiDAS object." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">midasHLA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.31</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/MiDAS_tutorial.html">MiDAS tutorial</a>
    </li>
    <li>
      <a href="../articles/MiDAS_vignette.html">MiDAS quick start</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Construct a MiDAS object</h1>
    
    <div class="hidden name"><code>prepareMiDAS.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>prepareMiDAS</code> transform HLA alleles calls and KIR calls according
to selected <code>experiment</code>s creating a <code><a href='MiDAS-class.html'>MiDAS</a></code> object.</p>
    </div>

    <pre class="usage"><span class='fu'>prepareMiDAS</span><span class='op'>(</span>
  hla_calls <span class='op'>=</span> <span class='cn'>NULL</span>,
  kir_calls <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>colData</span>,
  experiment <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"hla_alleles"</span>, <span class='st'>"hla_aa"</span>, <span class='st'>"hla_g_groups"</span>, <span class='st'>"hla_supertypes"</span>,
    <span class='st'>"hla_NK_ligands"</span>, <span class='st'>"kir_genes"</span>, <span class='st'>"kir_haplotypes"</span>, <span class='st'>"hla_kir_interactions"</span>,
    <span class='st'>"hla_divergence"</span>, <span class='st'>"hla_het"</span>, <span class='st'>"hla_custom"</span>, <span class='st'>"kir_custom"</span><span class='op'>)</span>,
  placeholder <span class='op'>=</span> <span class='st'>"term"</span>,
  lower_frequency_cutoff <span class='op'>=</span> <span class='cn'>NULL</span>,
  upper_frequency_cutoff <span class='op'>=</span> <span class='cn'>NULL</span>,
  indels <span class='op'>=</span> <span class='cn'>TRUE</span>,
  unkchar <span class='op'>=</span> <span class='cn'>FALSE</span>,
  hla_divergence_aa_selection <span class='op'>=</span> <span class='st'>"binding_groove"</span>,
  hla_het_resolution <span class='op'>=</span> <span class='fl'>8</span>,
  hla_dictionary <span class='op'>=</span> <span class='cn'>NULL</span>,
  kir_dictionary <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>hla_calls</th>
      <td><p>HLA calls data frame, as returned by
<code><a href='readHlaCalls.html'>readHlaCalls</a></code> function.</p></td>
    </tr>
    <tr>
      <th>kir_calls</th>
      <td><p>KIR calls data frame, as returned by
<code><a href='readKirCalls.html'>readKirCalls</a></code> function.</p></td>
    </tr>
    <tr>
      <th>colData</th>
      <td><p>Data frame holding additional variables like phenotypic
observations or covariates. It have to contain <code>'ID'</code> column holding
samples identifiers corresponding to identifiers in <code>hla_calls</code> and
<code>kir_calls</code>. Importantly rows of <code>hla_calls</code> and
<code>kir_calls</code> without corresponding phenotype are discarded.</p></td>
    </tr>
    <tr>
      <th>experiment</th>
      <td><p>Character vector indicating analysis type for which data
should be prepared. Valid choices are <code>"hla_alleles"</code>,
<code>"hla_aa"</code>, <code>"hla_g_groups"</code>, <code>"hla_supertypes"</code>,
<code>"hla_NK_ligands"</code>, <code>"kir_genes"</code>, <code>"hla_kir_interactions"</code>,
<code>"hla_divergence"</code>, <code>"hla_het"</code>.
See details for further explanations.</p></td>
    </tr>
    <tr>
      <th>placeholder</th>
      <td><p>String giving name for dummy variable inserted to
<code>colData</code>. This variable can be than used to define base statistical
model used by <code><a href='runMiDAS.html'>runMiDAS</a></code>.</p></td>
    </tr>
    <tr>
      <th>lower_frequency_cutoff</th>
      <td><p>Number giving lower frequency threshold.
Numbers greater than 1 are interpreted as the number of feature occurrences,
numbers between 0 and 1 as fractions.</p></td>
    </tr>
    <tr>
      <th>upper_frequency_cutoff</th>
      <td><p>Number giving upper frequency threshold.
Numbers greater than 1 are interpreted as the number of feature occurrences,
numbers between 0 and 1 as fractions.</p></td>
    </tr>
    <tr>
      <th>indels</th>
      <td><p>Logical indicating whether indels should be considered when
checking amino acid variability in <code>'hla_aa'</code> experiment.</p></td>
    </tr>
    <tr>
      <th>unkchar</th>
      <td><p>Logical indicating whether unknown characters in the alignment
should be considered when checking amino acid variability in
<code>'hla_aa'</code> experiment.</p></td>
    </tr>
    <tr>
      <th>hla_divergence_aa_selection</th>
      <td><p>String specifying variable region in peptide binding
groove which should be considered for Grantham distance calculation. Valid
choices includes: <code>"binding_groove"</code>, <code>"B_pocket"</code>,
<code>"F_pocket"</code>. See details for more information.</p></td>
    </tr>
    <tr>
      <th>hla_het_resolution</th>
      <td><p>Number specifying HLA alleles resolution used to
calculate heterogeneity in <code>"hla_het"</code> experiment.</p></td>
    </tr>
    <tr>
      <th>hla_dictionary</th>
      <td><p>Data frame giving HLA allele dictionary used in
<code>'hla_custom'</code> experiment. See <code><a href='hlaToVariable.html'>hlaToVariable</a></code> for more
details.</p></td>
    </tr>
    <tr>
      <th>kir_dictionary</th>
      <td><p>Data frame giving KIR genes dictionary used in
<code>'kir_custom'</code> experiment. See <code><a href='countsToVariables.html'>countsToVariables</a></code> for more
details.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Object of class <code><a href='MiDAS-class.html'>MiDAS</a></code></p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><code>experiment</code> specifies analysis types for which <code>hla_calls</code> and
<code>kir_call</code> should be prepared.</p><dl>
  <dt><code>'hla_alleles'</code></dt><dd><p><code>hla_calls</code> are transformed to counts matrix describing number of
    allele occurrences for each sample. This experiment is used to test
    associations on HLA alleles level.</p></dd>
  <dt><code>'hla_aa'</code></dt><dd><p><code>hla_calls</code> are transformed to a matrix of variable amino acid
    positions. See <code><a href='hlaToAAVariation.html'>hlaToAAVariation</a></code> for more details. This
    experiment is used to test associations on amino acid level.</p></dd>
  <dt><code>"hla_g_groups"</code></dt><dd><p><code>hla_calls</code> are translated into HLA G groups and transformed to
    matrix describing number of G group occurrences for each sample. See
    <code><a href='hlaToVariable.html'>hlaToVariable</a></code> for more details. This experiment is used to
    test associations on HLA G groups level.</p></dd>
  <dt><code>"hla_supertypes"</code></dt><dd><p><code>hla_calls</code> are translated into HLA supertypes and transformed to
    matrix describing number of G group occurrences for each sample. See
    <code><a href='hlaToVariable.html'>hlaToVariable</a></code> for more details. This experiment is used to
    test associations on HLA supertypes level.</p></dd>
  <dt><code>"hla_NK_ligands"</code></dt><dd><p><code>hla_calls</code> are translated into NK ligands, which includes HLA
    Bw4/Bw6 and HLA C1/C2 groups and transformed to matrix describing number
    of their occurrences for each sample. See <code><a href='hlaToVariable.html'>hlaToVariable</a></code> for
    more details.This experiment is used to test associations on HLA NK
    ligands level.</p></dd>
  <dt><code>"kir_genes"</code></dt><dd><p><code>kir_calls</code> are transformed to counts matrix describing number of
    KIR gene occurrences for each sample. This experiment is used to test
    associations on KIR genes level.</p></dd>
  <dt><code>"hla_kir_interactions"</code></dt><dd><p><code>hla_calls</code> and <code>kir_calls</code> are translated to HLA - KIR
    interactions as defined in
    <a href='https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6558367/'>Pende et al., 2019.</a>.
    See <code><a href='getHlaKirInteractions.html'>getHlaKirInteractions</a></code> for more details. This experiment
    is used to test associations on HLA - KIR interactions level.</p></dd>
  <dt><code>"hla_divergence"</code></dt><dd><p>Grantham distance for class I HLA alleles is calculated based on
    <code>hla_calls</code> using original formula by
    <a href='http://www.sciencemag.org/content/185/4154/862.long'>Grantham R. 1974.</a>.
    See <code><a href='hlaCallsGranthamDistance.html'>hlaCallsGranthamDistance</a></code> for more details. This
    experiment is used to test associations on HLA divergence level measured
    by Grantham distance.</p></dd>
  <dt><code>"hla_het"</code></dt><dd><p><code>hla_calls</code> are transformed to heterozygosity status, where <code>1</code>
    designates a heterozygote and <code>0</code> homozygote. Heterozygosity status
    is calculated only for classical HLA genes (A, B, C, DQA1, DQB1, DRA,
    DRB1, DPA1, DPB1). This experiment is used to test associations on HLA
    divergence level measured by heterozygosity.</p></dd>

</dl>


    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='va'>midas</span> <span class='op'>&lt;-</span> <span class='fu'>prepareMiDAS</span><span class='op'>(</span>hla_calls <span class='op'>=</span> <span class='va'>MiDAS_tut_HLA</span>,
                      kir_calls <span class='op'>=</span> <span class='va'>MiDAS_tut_KIR</span>,
                      colData <span class='op'>=</span> <span class='va'>MiDAS_tut_pheno</span>,
                      experiment <span class='op'>=</span> <span class='st'>"hla_alleles"</span><span class='op'>)</span>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Christian Hammer, Maciej Migdał.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


