<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Run MiDAS statistical analysis — runMiDAS • MiDAS</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Run MiDAS statistical analysis — runMiDAS" />
<meta property="og:description" content="runMiDAS perform association analysis on MiDAS data using statistical
model of choice. Function is intended for use with prepareMiDAS.
See examples section." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MiDAS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.13</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/MiDAS_tutorial.html">MiDAS tutorial</a>
    </li>
    <li>
      <a href="../articles/MiDAS_vignette.html">MiDAS quick start</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run MiDAS statistical analysis</h1>
    
    <div class="hidden name"><code>runMiDAS.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>runMiDAS</code> perform association analysis on MiDAS data using statistical
model of choice. Function is intended for use with <code><a href='prepareMiDAS.html'>prepareMiDAS</a></code>.
See examples section.</p>
    </div>

    <pre class="usage"><span class='fu'>runMiDAS</span><span class='op'>(</span>
  <span class='va'>object</span>,
  <span class='va'>experiment</span>,
  inheritance_model <span class='op'>=</span> <span class='cn'>NULL</span>,
  conditional <span class='op'>=</span> <span class='cn'>FALSE</span>,
  omnibus <span class='op'>=</span> <span class='cn'>FALSE</span>,
  omnibus_groups_filter <span class='op'>=</span> <span class='cn'>NULL</span>,
  lower_frequency_cutoff <span class='op'>=</span> <span class='cn'>NULL</span>,
  upper_frequency_cutoff <span class='op'>=</span> <span class='cn'>NULL</span>,
  correction <span class='op'>=</span> <span class='st'>"bonferroni"</span>,
  n_correction <span class='op'>=</span> <span class='cn'>NULL</span>,
  exponentiate <span class='op'>=</span> <span class='cn'>FALSE</span>,
  th <span class='op'>=</span> <span class='fl'>0.05</span>,
  th_adj <span class='op'>=</span> <span class='cn'>TRUE</span>,
  keep <span class='op'>=</span> <span class='cn'>FALSE</span>,
  rss_th <span class='op'>=</span> <span class='fl'>1e-07</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>An existing fit from a model function such as lm, glm and many
others.</p></td>
    </tr>
    <tr>
      <th>experiment</th>
      <td><p>String indicating the experiment associated with
<code>object</code>'s <code>MiDAS</code> data to use. Valid values includes:
<code>"hla_alleles"</code>, <code>"hla_aa"</code>, <code>"hla_g_groups"</code>,
<code>"hla_supertypes"</code>, <code>"hla_NK_ligands"</code>, <code>"kir_genes"</code>,
<code>"kir_haplotypes"</code>, <code>"hla_kir_interactions"</code>,
<code>"hla_divergence"</code>, <code>"hla_het"</code>, <code>"hla_custom"</code>,
<code>"kir_custom"</code>. See <code><a href='prepareMiDAS.html'>prepareMiDAS</a></code> for more information.</p></td>
    </tr>
    <tr>
      <th>inheritance_model</th>
      <td><p>String specifying inheritance model to use.
Available choices are <code>"dominant"</code>, <code>"recessive"</code>,
<code>"additive"</code>.</p></td>
    </tr>
    <tr>
      <th>conditional</th>
      <td><p>Logical flag indicating if conditional analysis should be
performed.</p></td>
    </tr>
    <tr>
      <th>omnibus</th>
      <td><p>Logical flag indicating if omnibus test should be used.</p></td>
    </tr>
    <tr>
      <th>omnibus_groups_filter</th>
      <td><p>Character vector specifying omnibus groups to
use.</p></td>
    </tr>
    <tr>
      <th>lower_frequency_cutoff</th>
      <td><p>Number giving lower frequency threshold.
Numbers greater than 1 are interpreted as the number of feature occurrences,
numbers between 0 and 1 as fractions.</p></td>
    </tr>
    <tr>
      <th>upper_frequency_cutoff</th>
      <td><p>Number giving upper frequency threshold.
Numbers greater than 1 are interpreted as the number of feature occurrences,
numbers between 0 and 1 as fractions.</p></td>
    </tr>
    <tr>
      <th>correction</th>
      <td><p>String specifying multiple testing correction method. See
details for further information.</p></td>
    </tr>
    <tr>
      <th>n_correction</th>
      <td><p>Integer specifying number of comparisons to consider
during multiple testing correction calculations. For Bonferroni correction
it is possible to specify a number lower than the number of comparisons
being made. This is useful in cases when knowledge about the biology or
redundance of alleles reduces the need for correction. For other methods it
must be at least equal to the number of comparisons being made; only set
this (to non-default) when you know what you are doing!</p></td>
    </tr>
    <tr>
      <th>exponentiate</th>
      <td><p>Logical flag indicating whether or not to exponentiate
the coefficient estimates. Internally this is passed to
<code><a href='https://generics.r-lib.org/reference/tidy.html'>tidy</a></code>. This is typical for logistic and multinomial
regressions, but a bad idea if there is no log or logit link. Defaults to
FALSE.</p></td>
    </tr>
    <tr>
      <th>th</th>
      <td><p>Number specifying threshold for a variable to be considered
significant.</p></td>
    </tr>
    <tr>
      <th>th_adj</th>
      <td><p>Logical flag indicating if adjusted p-value should be used as
threshold criteria, otherwise unadjusted p-value is used.</p></td>
    </tr>
    <tr>
      <th>keep</th>
      <td><p>Logical flag indicating if the output should be a list of results
resulting from each selection step. Default is to return only the final
result.</p></td>
    </tr>
    <tr>
      <th>rss_th</th>
      <td><p>Number specifying residual sum of squares threshold at which
function should stop adding additional variables. As the residual sum of
squares approaches <code>0</code> the perfect fit is obtained making further
attempts at variable selection nonsense. This behavior can be controlled
using <code>rss_th</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Analysis results, depending on the parameters:</p><dl>
    <dt><code>conditional=FALSE, omnibus=FALSE</code></dt><dd><p>Tibble with first column <code>"term"</code> holding names of tested 
      variables (eg. alleles). Further columns depends on the used 
      model and are determined by associated <code>tidy</code> function. Generally 
      they will include <code>"estimate"</code>, <code>"std.error"</code>, 
      <code>"statistic"</code>, <code>"p.value"</code>, <code>"conf.low"</code>, 
      <code>"conf.high"</code>, <code>"p.adjusted"</code>.</p></dd>
    <dt><code>conditional=TRUE, omnibus=FALSE</code></dt><dd><p>Tibble or a list of tibbles, see <code>keep</code> argument. The first column 
      <code>"term"</code> hold names of tested variables. Further 
      columns depends on the used model and are determined by associated 
      <code>tidy</code> function. Generally they will include <code>"estimate"</code>, 
      <code>"std.error"</code>, <code>"statistic"</code>, <code>"p.value"</code>, 
      <code>"conf.low"</code>, <code>"conf.high"</code>, <code>"p.adjusted"</code>.</p></dd>
    <dt><code>conditional=FALSE, omnibus=TRUE</code></dt><dd><p>Tibble with first column holding names of tested omnibus groups 
      (eg. amino acid positions) and second names of variables in the group
      (eg. residues). Further columns are: <code>"df"</code> giving difference in 
      degrees of freedom between base and extended model, <code>"statistic"</code>
      giving Chisq statistic, <code>"p.value"</code> and <code>"p.adjusted"</code>.</p></dd>
    <dt><code>conditional=TRUE, omnibus=TRUE</code></dt><dd><p>Tibble or a list of tibbles, see <code>keep</code> argument. The first column 
      hold names of tested omnibus groups (eg. amino acid positions), second 
      column hold names of variables in the group (eg. residues). Further 
      columns are: <code>"df"</code> giving difference in degrees of freedom 
      between base and extended model, <code>"statistic"</code> giving Chisq 
      statistic, <code>"p.value"</code> and <code>"p.adjusted"</code>.</p></dd>
  
</dl>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>By default statistical analysis is performed iteratively on each variable in
selected experiment. This is done by substituting <code>placeholder</code> in the
<code>object</code>'s formula with each variable in the experiment.</p>
<p>Setting <code>conditional</code> argument to <code>TRUE</code> will cause the statistical
analysis to be performed in a stepwise conditional testing manner, adding the
previous top-associated variable as a covariate to <code>object</code>'s formula.
The analysis stops when there is no more significant variables, based on
self-defined threshold (<code>th</code> argument). Either adjusted or unadjusted
p-values can be used as the selection criteria, which is controlled using
<code>th_adj</code> argument.</p>
<p>Setting <code>omnibus</code> argument to <code>TRUE</code> will cause the statistical
analysis to be performed iteratively on groups of variables (like residues at
particular amino acid position) using likelihood ratio test.</p>
<p>Argument <code>inheritance_model</code> specifies the inheritance model that should
be applyed to experiment's data. Following choices are available:</p><ul>
<li><p>"dominant"
    carrier status is sufficient for expression of the phenotype (non-carrier: 
    0, heterozygous &amp; homozygous carrier: 1).</p></li>
<li><p>"recessive"
    two copies are required for expression of the phenotype (non-carrier &amp; 
    heterozygous carrier: 0, homozygous carrier: 1).</p></li>
<li><p>"additive"
    allele dosage matters, homozygous carriers show stronger phenotype 
    expression or higher risk than heterozygous carriers (non-carrier = 0, 
    heterozygous carrier = 1, homozygous carrier = 2).</p></li>
<li><p>"overdominant"
    heterozygous carriers are at higher risk compared to non-carriers or 
    homozygous carriers (non-carrier &amp; homozygous carrier = 0, heterozygous 
    carrier = 1).</p></li>
</ul>  
 
<p><code>correction</code> specifies p-value adjustment method to use, common choice
is Benjamini &amp; Hochberg (1995) (<code>"BH"</code>). Internally this is passed to
<a href='https://rdrr.io/r/stats/p.adjust.html'>p.adjust</a>.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># create MiDAS object</span>
<span class='va'>midas</span> <span class='op'>&lt;-</span> <span class='fu'><a href='prepareMiDAS.html'>prepareMiDAS</a></span><span class='op'>(</span>hla_calls <span class='op'>=</span> <span class='va'>MiDAS_tut_HLA</span>,
                      colData <span class='op'>=</span> <span class='va'>MiDAS_tut_pheno</span>,
                      experiment <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"hla_alleles"</span>, <span class='st'>"hla_aa"</span><span class='op'>)</span>
<span class='op'>)</span>

<span class='co'># construct statistical model</span>
<span class='va'>object</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/lm.html'>lm</a></span><span class='op'>(</span><span class='va'>disease</span> <span class='op'>~</span> <span class='va'>term</span>, data <span class='op'>=</span> <span class='va'>midas</span><span class='op'>)</span>

<span class='co'># run analysis</span>
<span class='fu'>runMiDAS</span><span class='op'>(</span><span class='va'>object</span>, experiment <span class='op'>=</span> <span class='st'>"hla_alleles"</span>, inheritance_model <span class='op'>=</span> <span class='st'>"dominant"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span style='color: #555555;'># A tibble: 447 x 14</span><span>
#&gt;    allele p.value p.adjusted estimate std.error conf.low conf.high statistic
#&gt;    </span><span style='color: #555555;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span>      </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span>     </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span>     </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span>     </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span>
#&gt; </span><span style='color: #555555;'> 1</span><span> DQB1*… 1.73</span><span style='color: #555555;'>e</span><span style='color: #BB0000;'>-6</span><span>   0.000</span><span style='text-decoration: underline;'>773</span><span>   0.189     0.039</span><span style='text-decoration: underline;'>3</span><span>   0.112     0.266       4.81
#&gt; </span><span style='color: #555555;'> 2</span><span> DRB1*… 5.29</span><span style='color: #555555;'>e</span><span style='color: #BB0000;'>-6</span><span>   0.002</span><span style='text-decoration: underline;'>36</span><span>    0.177     0.038</span><span style='text-decoration: underline;'>7</span><span>   0.101     0.253       4.58
#&gt; </span><span style='color: #555555;'> 3</span><span> B*57:… 1.34</span><span style='color: #555555;'>e</span><span style='color: #BB0000;'>-5</span><span>   0.006</span><span style='text-decoration: underline;'>01</span><span>    0.254     0.058</span><span style='text-decoration: underline;'>1</span><span>   0.140     0.368       4.37
#&gt; </span><span style='color: #555555;'> 4</span><span> C*07:… 1.60</span><span style='color: #555555;'>e</span><span style='color: #BB0000;'>-4</span><span>   0.071</span><span style='text-decoration: underline;'>4</span><span>     0.148     0.039</span><span style='text-decoration: underline;'>0</span><span>   0.071</span><span style='text-decoration: underline;'>3</span><span>    0.224       3.79
#&gt; </span><span style='color: #555555;'> 5</span><span> B*18:… 4.93</span><span style='color: #555555;'>e</span><span style='color: #BB0000;'>-4</span><span>   0.220     -</span><span style='color: #BB0000;'>0.168</span><span>     0.048</span><span style='text-decoration: underline;'>1</span><span>  -</span><span style='color: #BB0000;'>0.262</span><span>    -</span><span style='color: #BB0000;'>0.073</span><span style='color: #BB0000;text-decoration: underline;'>7</span><span>     -</span><span style='color: #BB0000;'>3.50</span><span>
#&gt; </span><span style='color: #555555;'> 6</span><span> DRA*0… 2.63</span><span style='color: #555555;'>e</span><span style='color: #BB0000;'>-3</span><span>   1          0.096</span><span style='text-decoration: underline;'>0</span><span>    0.031</span><span style='text-decoration: underline;'>9</span><span>   0.033</span><span style='text-decoration: underline;'>5</span><span>    0.159       3.02
#&gt; </span><span style='color: #555555;'> 7</span><span> DQB1*… 4.34</span><span style='color: #555555;'>e</span><span style='color: #BB0000;'>-3</span><span>   1         -</span><span style='color: #BB0000;'>0.116</span><span>     0.040</span><span style='text-decoration: underline;'>6</span><span>  -</span><span style='color: #BB0000;'>0.196</span><span>    -</span><span style='color: #BB0000;'>0.036</span><span style='color: #BB0000;text-decoration: underline;'>4</span><span>     -</span><span style='color: #BB0000;'>2.86</span><span>
#&gt; </span><span style='color: #555555;'> 8</span><span> B*51:… 4.49</span><span style='color: #555555;'>e</span><span style='color: #BB0000;'>-3</span><span>   1          0.134     0.047</span><span style='text-decoration: underline;'>2</span><span>   0.041</span><span style='text-decoration: underline;'>8</span><span>    0.227       2.85
#&gt; </span><span style='color: #555555;'> 9</span><span> DRA*0… 1.02</span><span style='color: #555555;'>e</span><span style='color: #BB0000;'>-2</span><span>   1         -</span><span style='color: #BB0000;'>0.118</span><span>     0.045</span><span style='text-decoration: underline;'>7</span><span>  -</span><span style='color: #BB0000;'>0.207</span><span>    -</span><span style='color: #BB0000;'>0.027</span><span style='color: #BB0000;text-decoration: underline;'>9</span><span>     -</span><span style='color: #BB0000;'>2.57</span><span>
#&gt; </span><span style='color: #555555;'>10</span><span> DRB1*… 1.10</span><span style='color: #555555;'>e</span><span style='color: #BB0000;'>-2</span><span>   1         -</span><span style='color: #BB0000;'>0.404</span><span>     0.159   -</span><span style='color: #BB0000;'>0.715</span><span>    -</span><span style='color: #BB0000;'>0.092</span><span style='color: #BB0000;text-decoration: underline;'>9</span><span>     -</span><span style='color: #BB0000;'>2.55</span><span>
#&gt; </span><span style='color: #555555;'># … with 437 more rows, and 6 more variables: Ntotal </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #555555;'>,</span><span>
#&gt; </span><span style='color: #555555;'>#   Ntotal.percent </span><span style='color: #555555;font-style: italic;'>&lt;formttbl&gt;</span><span style='color: #555555;'>, `N(disease=0)` </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #555555;'>,</span><span>
#&gt; </span><span style='color: #555555;'>#   `N(disease=0).percent` </span><span style='color: #555555;font-style: italic;'>&lt;formttbl&gt;</span><span style='color: #555555;'>, `N(disease=1)` </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #555555;'>,</span><span>
#&gt; </span><span style='color: #555555;'>#   `N(disease=1).percent` </span><span style='color: #555555;font-style: italic;'>&lt;formttbl&gt;</span><span></div><div class='input'>
<span class='co'># omnibus test</span>
<span class='co'># omnibus_groups_filter argument can be used to restrict omnibus test only</span>
<span class='co'># to selected variables groups, here we restrict the analysis to HLA-A</span>
<span class='co'># positions 29 and 43.</span>
<span class='fu'>runMiDAS</span><span class='op'>(</span>
  <span class='va'>object</span>,
  experiment <span class='op'>=</span> <span class='st'>"hla_aa"</span>,
  inheritance_model <span class='op'>=</span> <span class='st'>"dominant"</span>,
  omnibus <span class='op'>=</span> <span class='cn'>TRUE</span>,
  omnibus_groups_filter <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"A_29"</span>, <span class='st'>"A_43"</span><span class='op'>)</span>
<span class='op'>)</span>
</div><div class='output co'>#&gt; </span><span style='color: #555555;'># A tibble: 2 x 6</span><span>
#&gt;   aa_pos residues    df statistic p.value p.adjusted
#&gt;   </span><span style='color: #555555;font-style: italic;'>&lt;chr&gt;</span><span>  </span><span style='color: #555555;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span>     </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span>   </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span>      </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span>
#&gt; </span><span style='color: #555555;'>1</span><span> A_43   Q, R         2      8.36  0.015</span><span style='text-decoration: underline;'>3</span><span>     0.030</span><span style='text-decoration: underline;'>6</span><span>
#&gt; </span><span style='color: #555555;'>2</span><span> A_29   D, A         1      1.00  0.317      0.634 </div><div class='input'>
</div></span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Christian Hammer, Maciej Migdał.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


